keps/sig-network/20190603-EndpointSlice-API.md

```diff
@@ -231,15 +239,23 @@ type EndpointPort struct {
-- **Endpoint Port**: The endpoint port number becomes optional in the EndpointSlice API while the port number field in core/v1 Endpoints API is required. This allows the API to support services with no port remapping or all port services.   
+- **Endpoint Port**: The endpoint port number becomes optional in the EndpointSlice API while the port number field in core/v1 Endpoints API is required. This allows the API to support services with no port remapping or all port services.
 
 ### Topology (Per Endpoint)
 A new topology field (string to string map) is added to each endpoint. It can contain arbitrary topology information associated with the endpoint. If the EndpointSlice instance is derived from K8s service, the topology may contain following well known key:
 - **kubernetes.io/hostname**: the value indicates the hostname of the node where the endpoint is located. This should match the corresponding node label.
 - **topology.kubernetes.io/zone**: the value indicates the zone where the endpoint is located. This should match the corresponding node label.
 - **topology.kubernetes.io/region**: the value indicates the region where the endpoint is located. This should match the corresponding node label.
 
-If the k8s service has topological keys specified, the corresponding node labels will be copied to endpoint topology. 
+If the k8s service has topological keys specified, the corresponding node labels will be copied to endpoint topology.
+
+### Termination Grace Period (Per Endpoint)
+
+An optional field `TerminationGracePeriodSeconds` can be added to each Endpoint indicating the longest duration an endpoint is given to gracefully shutdown. This is analogous to `TerminationGracePeriodSeconds` in `PodSpec`. In most cases, the endpointslice controller will set `TerminationGracePeriodSeconds` on an `Endpoint` based on the termination grace period on `PodSpec`. External implementations of the `EndpointSlice` API can set this field arbitrarily based on specific implementation requirements.
+
+This field allows consumers of the API to know the termination grace period of endpoints (pods in this case) without having to watch Pods. A specific use-case for this field is the [graceful termination handler in the IPVS proxy](https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/graceful_termination.go). Because deleting an IPVS backend (a.k.a IPVS real server) immediately deletes all its correpsonding conntrack entries, deleting a real server is disruptive during a pod's graceful termination period. Today, the IPVS proxy will only delete a real server if it sees that it has no connections or it has exceeded a hardcoded duration limit (15m). The `TerminationGracePeriodSeconds` field would remove the need for this hardcoded limit, improving overall performance of the proxy as it no longer needs to track stale backends.
+
+why was "not ready" addresses not used?
```

Comment:
This is a good question.  I don't think I know the answer.  @robscott you have more recent context... ?

Reply:
If NotReady endpoints were always in the slice, we could easily make the difference between a NotReady endpoint (weight=0) and a deleted endpoint (absent). However, I think NotReady endpoints are only present when PublishNotReadyAddresses is enabled.
If we always have NotReady endpoint in the slice, it will definitely be enough, but it would require some changes in the logic of kube-proxy (and maybe other tools relying on endpoints) to look at PublishNotReadyAddresses to decide on the behavior for a specific service

Reply:
Let's try to drag the conversation from PR back to here.

Ignore "tolerate unready", which just merges the two states.

Endpoints (not slice) puts pods in NotReadyAddresses on startup, but on teardown (when DeletionTimestamp is set) it removes them  entirely.

If we moved them BACK to NotReady on teardown it would be a little bit of a behavioral change, but it's semantically correct and seems unlikely to break users.  In case of tolerate-unready, we'd need to decide what the "right" behavior is.  Today pods being deleted are removed, regardless of tolerate-unready.  We could a) put them in NotReadyAddresses (which retains existing behavior -- they are not in Addresses) or b) put then in Addresses (which is a behavioral change, but consistent).

The bigger issue is whether this would cause a scalability problem.  It would cause another round of writes to the Endpoints object (ready->unready->gone vs today's ready->gone), which is already problematic at scale.  It seems that we don't actively batch updates to a single Endpoints.

For EndpointSlice we have the same question, though less pathologically.  For a single pod it's much less of a big deal.  In practice, we're likely to see this when a deployment is being updated, so it will hit all pods anyway.  This is why we held off adding more Conditions - each one is a cost scalar.  This is roughly equivalent to adding a new Condition (and indeed we should discuss that choice).

That said, I prefer making this sort of change to adding a new field.  I really want to keep EPSlice general-purpose and grace period feels weird.  Also, the above "memory loss" problem exists, whereas "unready" is more complete.

@robscott @wojtek-t re scalability.